main.cpp

```diff
@@ -328,325 +242,174 @@ struct FloatType
-    FloatType& pow(const DoubleType& exp);
-    FloatType& pow(float exp);
-
-private:
-    std::unique_ptr<float> value;
-    FloatType& powInternal(float exp);
-};
-
-FloatType& FloatType::operator+=(float rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator-=(float rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator*=(float rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator/=(float rhs)
-{
-    if (rhs == 0.f)
+    Numeric& operator+=(Type rhs)
     {
-        std::cout << "warning: floating point division by zero!" << std::endl;
+        *value += rhs;
+        return *this;
     }
 
-    *value /= rhs;
-    return *this;
-}
-
-/*
-DOUBLE
-*/
-struct DoubleType
-{
-    explicit DoubleType(double doubleValue) : value( std::make_unique<double>(doubleValue) ) { }
+    Numeric& operator-=(Type rhs)
+    {
+        *value -= rhs;
+        return *this;
+    }
 
-    DoubleType& apply(std::function<DoubleType&( std::unique_ptr<double>& )> func)
+    Numeric& operator*=(Type rhs)
     {
-        if( func )
-        {
-            func(value);
-        }
+        *value *= rhs;
         return *this;
     }
 
-    using DoubleFunctionPointer = void(*)( std::unique_ptr<double>& );
-    DoubleType& apply( DoubleFunctionPointer functionPtr )
+    template<typename divideByType>
+    Numeric& operator/=(divideByType&& rhs)
     {
-        if( functionPtr )
+        if constexpr (std::is_same<int, Type>::value)
         {
-            functionPtr(value);
+            if constexpr (std::is_same<int, divideByType>::value)
+            {
+                if (rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!\n";
+                    return *this;
+                }
+            }
+            else if (rhs < std::numeric_limits<divideByType>::epsilon() )
+            {
+                std::cout << "can't divide integers by zero!\n";
+                return *this;
+            }
         }
+        else if (rhs < std::numeric_limits<Type>::epsilon() )
+        {
+            std::cout << "warning: floating point division by zero!\n";
+        }
+
+        *value /= rhs;
         return *this;
     }
 
-    operator double() const { return *value; }
-
-    DoubleType& operator+=(double rhs);
-    DoubleType& operator-=(double rhs);
-    DoubleType& operator*=(double rhs);
-    DoubleType& operator/=(double rhs);
-
-    DoubleType& pow(const IntType& exp);
-    DoubleType& pow(const FloatType& exp);
-    DoubleType& pow(const DoubleType& exp);
-    DoubleType& pow(double exp);
+    Numeric& pow(const Numeric<float>& exp) { return powInternal( static_cast<Type>(exp) ); }
+    Numeric& pow(const Numeric<int>& exp) { return powInternal( static_cast<Type>(exp) ); }
+    Numeric& pow(const Numeric<double>& exp) { return powInternal( static_cast<Type>(exp) ); }
```

Comment:
Comment these out.  see if they're even needed anymore.
If everything works as expected, you can delete these functions entirely
- [ ] test and revise